ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -12,17 +25,466 @@ pub trait DZKPCompatibleField: Field {
+    // We use the conversion logic from https://eprint.iacr.org/2023/909.pdf
+    //
+    // This function does not use any optimization.
+    //
+    // Prover and left can compute:
+    // g1=-2ac(1-2e),
+    // g2=c(1-2e),
+    // g3=a(1-2e),
+    // g4=-1/2(1-2e),
+    //
+    // Prover and right can compute:
+    // h1=bd(1-2f),
+    // h2=d(1-2f),
+    // h3=b(1-2f),
+    // h4=1-2f,
+    //
+    // where
+    // (a,b,c,d,f) = (x_left, y_right, y_left, x_right, prss_right)
+    // and
+    // e = ab⊕cd⊕ f
+    // e is defined different from the paper here,
+    // in the paper e is defined as e = x_left · y_left ⊕ z_left ⊕ prss_left
+    // notice that the prover is supposed to show that ab⊕cd⊕ f⊕ e =0
+    // therefore e = ab⊕cd⊕ f must hold. (alternatively, you can also see this by substituting z_left,
+    // i.e. z_left = x_left · y_left ⊕ x_left · y_right ⊕ x_right · y_left ⊕ prss_left ⊕ prss_right
+    fn convert_prover<'a>(
+        x_left: &'a BitArray32,
+        x_right: &'a BitArray32,
+        y_left: &'a BitArray32,
+        y_right: &'a BitArray32,
+        prss_right: &'a BitArray32,
+    ) -> impl Iterator<Item = UVPolynomials<Fp61BitPrime>> {
+        x_left
+            .chunks(RECURSION_CHUNK_SIZE_BITS)
+            .zip(y_right.chunks(RECURSION_CHUNK_SIZE_BITS))
+            .zip(y_left.chunks(RECURSION_CHUNK_SIZE_BITS))
+            .zip(x_right.chunks(RECURSION_CHUNK_SIZE_BITS))
+            .zip(prss_right.chunks(RECURSION_CHUNK_SIZE_BITS))
+            .map(|((((a, b), c), d), f)| {
+                // e = ab⊕cd⊕ f = x_left * y_right ⊕ y_left * x_right ⊕ prss_right
+                let e: BitArray<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0> =
+                    (BitArray::try_from(a).unwrap()
+                        & BitArray::<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0>::try_from(b)
+                            .unwrap())
+                        ^ (BitArray::<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0>::try_from(c)
+                            .unwrap()
+                            & BitArray::<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0>::try_from(d)
+                                .unwrap())
+                        ^ BitArray::<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0>::try_from(f)
+                            .unwrap();
+                // precompute ac = a & c
+                let ac: BitArray<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0> =
+                    BitArray::try_from(a).unwrap()
+                        & BitArray::<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0>::try_from(c)
+                            .unwrap();
+                // precompute bd = b & d
+                let bd: BitArray<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0> =
+                    BitArray::try_from(b).unwrap()
+                        & BitArray::<[u8; RECURSION_CHUNK_SIZE_BITS >> 3], Lsb0>::try_from(d)
+                            .unwrap();
+                (
```

Comment:
I want to find a better solution that does not involve casing to BitArray. I think this code is just too difficult to read and understand, despite the fact that what's going on is really simple. I'm afraid it's going to make this hard to maintain, and hard for new readers to ramp up on. Let's find a way to just use native types and bitwise operations like '|' and '&'. I think we should consider a different storage type if the current storage type is forcing us to do this.

Reply:
I have changed it so that it is more readable.